sim/src/main/cc/endpoints/uart.cc

```diff
@@ -24,23 +33,53 @@ void sighand(int s) {
 }
 #endif
 
-uart_t::uart_t(simif_t* sim, UARTWIDGET_struct * mmio_addrs): endpoint_t(sim)
+uart_t::uart_t(simif_t* sim, UARTWIDGET_struct * mmio_addrs, int uartno): endpoint_t(sim)
 {
     this->mmio_addrs = mmio_addrs;
+    this->loggingfd = 0; // unused
 
-    // Don't block on stdin reads if there is nothing typed in
-    fcntl(STDIN_FILENO, F_SETFL, fcntl(STDIN_FILENO, F_GETFL) | O_NONBLOCK);
+    if (uartno == 0) {
+        // signal handler so ctrl-c doesn't kill simulation when UART is attached
+        // to stdin/stdout
+        struct sigaction sigIntHandler;
+        sigIntHandler.sa_handler = sighand;
+        sigemptyset(&sigIntHandler.sa_mask);
+        sigIntHandler.sa_flags = 0;
+        sigaction(SIGINT, &sigIntHandler, NULL);
+        printf("UART0 is here (stdin/stdout).\n");
+        inputfd = STDIN_FILENO;
+        outputfd = STDOUT_FILENO;
+    } else {
+        // for UARTs that are not UART0, use a PTY
+        char slavename[SLAVENAMELEN];
+        int ptyfd = open("/dev/ptmx", O_RDWR);
```

Comment:
`posix_openpt(O_RDWR | O_NOCTTY)` is a portable way to obtain the pty master file descriptor, as not all platforms have `/dev/ptmx` (although not too relevant given that firesim is tied to CentOS).